### DISCUSSION TITLE
Hitzetod!?

### DISCUSSION THREAD

**Discussion by Omega13x** [jomjol/AI-on-the-edge-device#2635] [General]
Hallo,
ich habe es auch geschafft das AI on the edge in Betrieb zu nehmen. Aber mein ESP wird sehr heiß. Zuerst hatte ich zwei Noname ESP getestet. Die sind nach kurzer Zeit mit Temperaturen von >100°C abgeraucht und die SD-Karten gleich mit. Jetzt habe ich ein ESP32-Cam von AZ-Delivery in Betrieb. Aber auch der wird trotz Kühlkörper manchmal über 85°C heiß. Der ESP ist offen verbaut. Kann es sein, das der ESP32 an der Grenze seiner Leistung ist? Hat noch wer die gleiche Erfahrung gemacht?

**Comment by friedpa** [comment]
Ich habe auch zwei ESP32 von AZ-Delivery im Betrieb, diese überschreiten aber nie 55°C.

**Reply by Omega13x** [reply]
AZ-Delivery haben mir zugesagt einen neuen kostenlos zu schicken, was ganz toll und zuvorkommend ist. Mal sehen wie der sich verhält. Am 0,3l Getränkebecher, den ich als Halterung verwende, kann es ja nicht liegen. 😉

**Comment by caco3** [comment]
Überhitzung war bis jetzt nie ein Problem.

Die Temperatur-Ausgabe ist jedoch nicht sehr vertrauenswürdig, insbesondere on rev. 3 boards!

**Comment by Omega13x** [comment]
Heute ist mir der dritte ESP32-Cam abgeraucht. Letzte Ausgabe war 113°C. Die SD-Karte ist natürlich auch hinüber. Ich verstehe das nicht, der ESP liegt offen auf dem Halter, Raumtemperatur liegt, wegen der Heizung und Solarthermie, so bei 25°C. Wenn ich die Ersatzlieferung von AZ-Delivery die nächsten Tage bekomme, hänge ich mal einen kleinen Lüfter über die Konstruktion. Wenn der auch abraucht, gebe ich auf. 

![20230930_195222](https://github.com/jomjol/AI-on-the-edge-device/assets/123419851/6d751646-e3fc-4f0c-badd-77da73f50c1d)

**Reply by caco3** [reply]
Woher weisst Du, dass sie wegen der Temperatur zerstört wurden? Und wieso sollte das auch der SD-Karte schaden?
Ich würde mal deine Speisung anschauen, sind das wirklich 5V?

**Reply by Omega13x** [reply]
Ich nutze ein Raspberry Netzteil mit 5,1V 2,5A. Die Spannung ist stabil bis >2A. Daran kann es nicht liegen. Die SD-Karten sind definitiv kaputt. Sie werden in keinem Gerät mehr erkannt. Ich bin aber wohl der Einzige mit diesem Problem.

**Reply by caco3** [reply]
Ja, die Speisung sollte ok sein.
Aber auch wenn die CPU wegen der Temperatur stirbt, sollte das keinen Einfluss auf die SD-Karte haben!

**Reply by karstengit** [reply]
Hast Du Mal den Betrieb einer anderen (Demo) Software probiert und geschaut wie es sich da verhält?
Bei den (nicht gerade geringen) "üblichen" 53°C die im Betrieb dieser Software auftreten, wird die Abschirmung bereits gut warm.
Dann dürfte ja bei 113°C die Abschirmung so heiß werden, daß man einen Finger nicht mehr drauf halten kann?

Ich hatte [hier](https://github.com/jomjol/AI-on-the-edge-device/discussions/2629#discussioncomment-7132767) ebenfalls das Problem das eine SD-Karte zumindest von dieser Software plötzlich nicht mehr gelesen wurde.
Woran das liegt kann ich nicht sagen, denn die SD-Karte läßt sich unter Linux problemlos auf dem PC lesen.

**Reply by caco3** [reply]
> Woran das liegt kann ich nicht sagen, denn die SD-Karte läßt sich unter Linux problemlos auf dem PC lesen.

Der von ESP-IDF bereitgestellte SDCard-Driver scheint sehr picky zu sein. Der verwirft vermutlich bei der kleinsten Abnormalität die Hände. Linux/Windows wird da viel Toleranter sein. Suche mal in den Diskussionen  nach Formatierungen der SD-Karte.

**Reply by karstengit** [reply]
> Der von ESP-IDF bereitgestellte SDCard-Driver scheint sehr picky zu sein.

Warum verwendet man dann nicht einfach eine andere Library?
Mit dieser hier habe ich gute Erfahrungen gemacht - sie läuft sehr stabil mit allen meinen Data-Loggern.
https://github.com/greiman/SdFat
Allerdings habe ich diese noch nicht mit dem ESP benutzt, sondern mit STM32.
Das sollte hinsichtlich der Benutzung von SPI aber keine wirklichen Unterschiede ausmachen?

**Reply by caco3** [reply]
Weil es ein Hobby-Projekt ist und wir limitierte Ressourcen haben.

Du darfst gerne einen Pull Request mit dem neuen Treiber einreichen. Da wir die SD-Karte aber recht intensiv nutzen, ist es fraglich, ob es wirklich stabiler wäre.

**Reply by Omega13x** [reply]
@karstengit
Ja, ich habe noch eine Kamera mit Tasmota laufen. Die wird nur leicht warm. Das Infrarot Thermometer zeigt zwischen 30 und 45Grad an.

**Comment by friedpa** [comment]
Ein Klebeband über den Kühlkörpern ist kontraproduktiv für die Hitzeableitung.

**Reply by Omega13x** [reply]
Das Klebeband berührt den Kühlkörper nicht. Sind noch gut zwei Millimeter platz. Normal sollte das Ganze auch ohne Kühlkörper zuverlässig funktionieren.

**Reply by karstengit** [reply]
Es ist die gleiche ESP32 CAM die mit der AI-Software deutlich heißer wird?

**Reply by Omega13x** [reply]
Das war mein erster Aufbau mit einem ESP von AZ-Delivery. Dieser ESP32-Cam lebt noch. Den habe ich, als die Temperatur über 80°C gestiegen ist, sofort abgeschaltet. Danach habe ich mit verschiedenen Noname-Chinesen getestet. Die sind aber jetzt alle drei kaputt. Hat sich auch erledigt. Mein Wasserversorger hat angekündigt, das ab 1. November alle Wasserzähler in unserem Dorf gegen digitale Geräte ausgetauscht werden.

**Reply by karstengit** [reply]
Hmm - kannst Du bitte etwas genauer schildern warum alle drei "Noname-Chinesen" kaputt sind und was der Unterschied zu der ESP-Cam von AZ-Delivery ist?
Letzten Endes ist das alles "Made in China" - vor allem das Kern-Modul ESP32 wird von allen wahrscheinlich aus der gleichen Quelle eingekauft.

**Reply by Omega13x** [reply]
Ich denke, die Noname Module sind kaputt gegangen weil ich sie nicht abgeschaltet habe. Sie haben nur halb so viel gekostet wie das Modul von AZ-Delivery.

**Reply by karstengit** [reply]
Warum sollten die Module keinen Dauerbetrieb aushalten?
Es ist allerdings merkwürdig das die ESP32 so warm werden.
Ich habe den Verdacht das es daran liegt das die Wärme einfach nicht gut abgeführt wird.
Auf der Rückseite der Adapterplatine ist der Stecker für die Kamera und somit keine Kühlfläche wo die CPU ist.

Allerdings muß ich zugestehen das ich bislang nur ESP8266 im Dauerbetrieb habe und die zudem ohne WLAN.
(WLAN wird optional aktiviert um Daten aus LittleFS per FTP abzuholen oder zu manipulieren.)
Die ESP laufen sehr stabil und sind sehr sparsam (20mA) und dennoch leistungsfähig und die integrierten 3 MB Flash als Datenspeicher für Datenlogger reichen für viele Zwecke.

**Reply by caco3** [reply]
Ich habe 2 ESP32-CAMs von Aliexpress, und die laufen seit 1 Jahr ohne Probleme (zeigen ca. 50°C).
Und auch andere ESP332 und ESP8622 von Aliexpress, und die laufen seit Jahren ohne Probleme.

**Reply by karstengit** [reply]
Die Qualität hat leider eine sehr große Streuung und es ist nicht vorhersehbar was man bekommt.
Wenn ein Produkt gut / brauchbar ist, hilft es möglichst schnell beim gleichen Händler noch ein paar mehr zu bestellen.
Dann ist die Wahrscheinlichkeit noch recht groß welche aus der gleichen Charge zu erhalten.

**Comment by karstengit** [comment]
> Weil es ein Hobby-Projekt ist und wir limitierte Ressourcen haben.
> 
> Du darfst gerne einen Pull Request mit dem neuen Treiber einreichen. Da wir die SD-Karte aber recht intensiv nutzen, ist es fraglich, ob es wirklich stabiler wäre.

Wie Du [in dem anderen Thread](https://github.com/jomjol/AI-on-the-edge-device/discussions/2629) sehen kannst, besteht für mich zur Zeit noch nicht der Durchblick in dem mehr oder weniger undokumentierten Code. Insbesondere die espidf Umgebung bereitet hier unnötig viele Einstiegs-Probleme.

Ich würde am liebsten auf die SD-Karte verzichten und diese nur für das Setup der config.ini benutzen.
Es ist noch nicht einmal klar, ob das Bild vor der Auswertung auf der SD-Karte zwischengespeichert wird, oder im PSRAM abgelegt und ausgewertet wird?
Die Konfiguration erlaubt eine Sicherung des Bildes, das hat aber genauso wie das extensive Logging nicht zwingend mit der Bilderkennung zu tun.

**Reply by caco3** [reply]
@Slider0007 hat dir da ja schon aufgezeigt, für was die SD-Karte verwendet wird. Wir können nicht auf die SD-Karte verzichten, weil  die Modelle zu gross sind und nicht direkt in der Firmware platz haben.

**Reply by karstengit** [reply]
Im Betrieb nach der Konfiguration wird doch nur ein Modell oder maximal zwei Modelle benötigt. Im Extremfall
Model = /config/dig-class11_1600_s2.tflite **348,1 KB**
Model = /config/ana-class100_0170_s1_q.tflite **130,5 KB**
(Woraus bestehen eigentlich diese Dateien? Sind das sowas wie Regelsätze oder ist das kompilierter Code?)

Beide zusammen benötigen max. 500 KB und könnten in einem LittleFS abgelegt werden, sofern sich LittleFS im espidf verwenden läßt?
Es sollten doch noch 1 MB Flash hinter der Firmware frei sein?

![partitions csv](https://github.com/jomjol/AI-on-the-edge-device/assets/5941550/098c3479-1801-4a8f-a335-6eae419a0e3b)

**Reply by caco3** [reply]
Das grösste Model belegt 1.1. MB:
![grafik](https://github.com/jomjol/AI-on-the-edge-device/assets/1783586/7abc150c-c5c6-46f0-8603-75a90def0a8e)

Momentan stehen wir mit 75% Belegung gut da. D.h. es sind so ca. 500 kB verfügbar:
```
RAM:   [==        ]  15.3% (used 50240 bytes from 327680 bytes)
Flash: [========  ]  75.6% (used 1471765 bytes from 1945600 bytes)
```
Früher war das nicht so, da kratzten wir sogar an 98%! Erst in den letzten Monaten könnten wir endlich massiv Platz schaffen, indem wir neuere Bibliotheken eingebunden haben, welche modularer sind.

Denk daran, dass wir unser Projekte generisch halten wollen, d.h., wir müssen verschiedene Modelle anbieten. Es hält dich aber niemand davon ab, das ganze für dich umzubauen und die Modelle direkt zu integrieren.

Diese Modelle sind kompilierte tensorflow-Modelle (Neuronale Netzwerke), welche aus den Bildern von Digits und Zeigern generiert wurden. Siehe https://jomjol.github.io/AI-on-the-edge-device-docs/Learn-models-with-your-own-images/ und andere Stellen in der Doku.

**Reply by karstengit** [reply]
Warum sind Deine Dateien größer?
In dem Ordner ../sd-card/config/ ist bei mir

```
131K  ana-class100_0170_s1_q.tflite
 47K  ana-cont_1206_s2_q.tflite
 47K  ana-cont_1207_s2_q.tflite
2,3K  config.ini
222K  dig-class100-0167_s2_q.tflite
349K  dig-class11_1600_s2.tflite
309K  dig-cont_0611_s3_q.tflite
309K  dig-cont_0620_s3_q.tflite
  30  prevalue.ini
2,2K  ref0.jpg
1,7K  ref0_org.jpg
1,8K  ref1.jpg
1,6K  ref1_org.jpg
109K  reference.jpg
```

Ist dies ein eigenes tensorflow-Modell von Dir?

Was mir bislang noch nicht ganz klar wurde ist die Benutzung des verbauten Flash von 4 MB:

```
Chip is ESP32-D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
Uploading stub...
Running stub...
Stub running...
Manufacturer: 5e
Device: 4016
Detected flash size: 4MB
```
Dieses Video erklärt es gut: https://www.youtube.com/watch?v=Qu-1RK4Fk7g

![partition](https://github.com/jomjol/AI-on-the-edge-device/assets/5941550/f6ef0cfa-57ee-4145-9fae-d18e8e326062)

Damit ist das Thema LittleFS vom Tisch. :-)

> Denk daran, dass wir unser Projekte generisch halten wollen, d.h., wir müssen verschiedene Modelle anbieten. Es hält dich aber niemand davon ab, das ganze für dich umzubauen und die Modelle direkt zu integrieren.

Ja - es geht mir nicht darum das ganze Projekt umzukrempeln, sondern lediglich um die [beschriebenen Ideen](https://github.com/jomjol/AI-on-the-edge-device/discussions/2629).
Ehrlich gesagt geht es mir darum mein Heim nicht in eine WLAN-Funkhölle zu verwandeln, vor allem da ohnehin Kabel für die Spannungsversorgung erforderlich sind. Bei mir sind alle Sensoren über einen eigenen Haus-Bus mit RS-485 vernetzt.
Vielleicht gelingt es mir ja die UART wie von mir vorgeschlagen zu nutzen - das braucht bei mir Zeit ...
